File: Components/WinForms/CS/ChartsMainDemo/Modules/DataAnalysis/MovingAverage.cs
using System;
using System.Collections.Generic;
using System.Globalization;

namespace DevExpress.XtraCharts.Demos.DataAnalysis {
    public partial class MovingAverageDemo : ChartDemoBase2D {
        ChartControl chart;

        List<MovingAverage> MovingAverageIndicators {
            get {
                if (Series == null)
                    return null;
                XYDiagram2DSeriesViewBase view = Series.View as XYDiagram2DSeriesViewBase;
                if (view == null)
                    return null;
                List<MovingAverage> movingAverageIndicators = new List<MovingAverage>();
                foreach (Indicator indicator in view.Indicators) {
                    MovingAverage movingAverage = indicator as MovingAverage;
                    if (movingAverage != null)
                        movingAverageIndicators.Add(movingAverage);
                }
                if (movingAverageIndicators != null)
                    return movingAverageIndicators;
                return null;
            }
        }

        public override ChartControl ChartControl { get { return this.chart; } }
        Series Series { get { return ChartControl.Series.Count > 0 ? ChartControl.Series[0] : null; } }

		public MovingAverageDemo() {
			InitializeComponent();
		}

        void SetVisibleIndicator() {
            if(MovingAverageIndicators == null)
                return;
            switch ((string)cbType.SelectedItem) {
                case "Simple":
                    foreach (MovingAverage indicator in MovingAverageIndicators)
                        indicator.Visible = indicator is SimpleMovingAverage;
                    break;
                case "Exponential":
                    foreach (MovingAverage indicator in MovingAverageIndicators)
                        indicator.Visible = indicator is ExponentialMovingAverage;
                    break;
                case "Weighted":
                    foreach (MovingAverage indicator in MovingAverageIndicators)
                        indicator.Visible = indicator is WeightedMovingAverage;
                    break;
                case "Triangular":
                    foreach (MovingAverage indicator in MovingAverageIndicators)
                        indicator.Visible = indicator is TriangularMovingAverage;
                    break;
                case "Triple Exponential (TEMA)":
                    foreach (MovingAverage indicator in MovingAverageIndicators)
                        indicator.Visible = indicator is TripleExponentialMovingAverageTema;
                    break;
                default:
                    foreach (MovingAverage indicator in MovingAverageIndicators)
                        indicator.Visible = indicator is SimpleMovingAverage;
                    break;
            }
        }
        void cbType_SelectedIndexChanged(object sender, EventArgs e) {
            SetVisibleIndicator();
        }
        void spnPointsCount_EditValueChanged(object sender, EventArgs e) {
            if(MovingAverageIndicators == null)
                return;
            foreach (MovingAverage indicator in MovingAverageIndicators)
                indicator.PointsCount = Convert.ToInt32(spnPointsCount.EditValue);
        }
        void cbKind_SelectedIndexChanged(object sender, EventArgs e) {
            if(MovingAverageIndicators == null)
                return;
            foreach (MovingAverage indicator in MovingAverageIndicators)
                indicator.Kind = (MovingAverageKind)cbKind.SelectedIndex;
            UpdateControls();
        }
        void spnEnvelopPercent_EditValueChanged(object sender, EventArgs e) {
            if(MovingAverageIndicators == null)
                return;
            foreach (MovingAverage indicator in MovingAverageIndicators)
                indicator.EnvelopePercent = Convert.ToDouble(spnEnvelopPercent.EditValue);
        }
        void chart_BoundDataChanged(object sender, EventArgs e) {
            if (Series != null && Series.Points.Count > 0) {
                spnPointsCount.Properties.MaxValue = Series.Points.Count;
                if (MovingAverageIndicators.Count > 0)
                    if (MovingAverageIndicators[0] != null) {
                        cbKind.SelectedIndex = (int)MovingAverageIndicators[0].Kind;
                        spnPointsCount.EditValue = MovingAverageIndicators[0].PointsCount;
                        spnEnvelopPercent.EditValue = MovingAverageIndicators[0].EnvelopePercent;
                    }
            }
        }

        protected override void InitControls() {
            base.InitControls();
            ShowLabels = false;
            Series.DataSource = MarketPrices.GetGoogleStockPrices();
        }

        public override void UpdateControls() {
            base.UpdateControls();
            SetVisibleIndicator();
            bool enableEnvelopPercent = cbKind.SelectedItem.ToString() != "Moving Average";
            lblEnvelopPercent.Enabled = enableEnvelopPercent;
            spnEnvelopPercent.Enabled = enableEnvelopPercent;
        }

        private void lblKind_Click(object sender, EventArgs e) {

        }
    }
}

File: Components/WinForms/VB/ChartsMainDemo/Modules/DataAnalysis/MovingAverage.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Globalization

Namespace DevExpress.XtraCharts.Demos.DataAnalysis
	Partial Public Class MovingAverageDemo
		Inherits ChartDemoBase2D
		Private WithEvents chart As ChartControl

		Private ReadOnly Property MovingAverageIndicators() As List(Of MovingAverage)
			Get
				If Series Is Nothing Then
					Return Nothing
				End If
				Dim view As XYDiagram2DSeriesViewBase = TryCast(Series.View, XYDiagram2DSeriesViewBase)
				If view Is Nothing Then
					Return Nothing
				End If
				Dim movingAverageIndicators_Renamed As New List(Of MovingAverage)()
				For Each indicator As Indicator In view.Indicators
					Dim movingAverage As MovingAverage = TryCast(indicator, MovingAverage)
					If movingAverage IsNot Nothing Then
						movingAverageIndicators_Renamed.Add(movingAverage)
					End If
				Next indicator
				If movingAverageIndicators_Renamed IsNot Nothing Then
					Return movingAverageIndicators_Renamed
				End If
				Return Nothing
			End Get
		End Property

		Public Overrides ReadOnly Property ChartControl() As ChartControl
			Get
				Return Me.chart
			End Get
		End Property
		Private ReadOnly Property Series() As Series
			Get
				Return If(ChartControl.Series.Count > 0, ChartControl.Series(0), Nothing)
			End Get
		End Property

		Public Sub New()
			InitializeComponent()
		End Sub

		Private Sub SetVisibleIndicator()
			If MovingAverageIndicators Is Nothing Then
				Return
			End If
			Select Case CStr(cbType.SelectedItem)
				Case "Simple"
					For Each indicator As MovingAverage In MovingAverageIndicators
						indicator.Visible = TypeOf indicator Is SimpleMovingAverage
					Next indicator
				Case "Exponential"
					For Each indicator As MovingAverage In MovingAverageIndicators
						indicator.Visible = TypeOf indicator Is ExponentialMovingAverage
					Next indicator
				Case "Weighted"
					For Each indicator As MovingAverage In MovingAverageIndicators
						indicator.Visible = TypeOf indicator Is WeightedMovingAverage
					Next indicator
				Case "Triangular"
					For Each indicator As MovingAverage In MovingAverageIndicators
						indicator.Visible = TypeOf indicator Is TriangularMovingAverage
					Next indicator
				Case "Triple Exponential (TEMA)"
					For Each indicator As MovingAverage In MovingAverageIndicators
						indicator.Visible = TypeOf indicator Is TripleExponentialMovingAverageTema
					Next indicator
				Case Else
					For Each indicator As MovingAverage In MovingAverageIndicators
						indicator.Visible = TypeOf indicator Is SimpleMovingAverage
					Next indicator
			End Select
		End Sub
		Private Sub cbType_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbType.SelectedIndexChanged
			SetVisibleIndicator()
		End Sub
		Private Sub spnPointsCount_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs) Handles spnPointsCount.EditValueChanged
			If MovingAverageIndicators Is Nothing Then
				Return
			End If
			For Each indicator As MovingAverage In MovingAverageIndicators
				indicator.PointsCount = Convert.ToInt32(spnPointsCount.EditValue)
			Next indicator
		End Sub
		Private Sub cbKind_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbKind.SelectedIndexChanged
			If MovingAverageIndicators Is Nothing Then
				Return
			End If
			For Each indicator As MovingAverage In MovingAverageIndicators
				indicator.Kind = CType(cbKind.SelectedIndex, MovingAverageKind)
			Next indicator
			UpdateControls()
		End Sub
		Private Sub spnEnvelopPercent_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs) Handles spnEnvelopPercent.EditValueChanged
			If MovingAverageIndicators Is Nothing Then
				Return
			End If
			For Each indicator As MovingAverage In MovingAverageIndicators
				indicator.EnvelopePercent = Convert.ToDouble(spnEnvelopPercent.EditValue)
			Next indicator
		End Sub
		Private Sub chart_BoundDataChanged(ByVal sender As Object, ByVal e As EventArgs) Handles chart.BoundDataChanged
			If Series IsNot Nothing AndAlso Series.Points.Count > 0 Then
				spnPointsCount.Properties.MaxValue = Series.Points.Count
				If MovingAverageIndicators.Count > 0 Then
					If MovingAverageIndicators(0) IsNot Nothing Then
						cbKind.SelectedIndex = CInt(Fix(MovingAverageIndicators(0).Kind))
						spnPointsCount.EditValue = MovingAverageIndicators(0).PointsCount
						spnEnvelopPercent.EditValue = MovingAverageIndicators(0).EnvelopePercent
					End If
				End If
			End If
		End Sub

		Protected Overrides Sub InitControls()
			MyBase.InitControls()
			ShowLabels = False
			Series.DataSource = MarketPrices.GetGoogleStockPrices()
		End Sub

		Public Overrides Sub UpdateControls()
			MyBase.UpdateControls()
			SetVisibleIndicator()
			Dim enableEnvelopPercent As Boolean = cbKind.SelectedItem.ToString() <> "Moving Average"
			lblEnvelopPercent.Enabled = enableEnvelopPercent
			spnEnvelopPercent.Enabled = enableEnvelopPercent
		End Sub

		Private Sub lblKind_Click(ByVal sender As Object, ByVal e As EventArgs) Handles lblKind.Click

		End Sub
	End Class
End Namespace
